File: src/di/ServiceContainer.ts
import { MockVitalSignProvider } from '../data/datasources/MockVitalSignProvider';
import { SymptomLocalDataSource } from '../data/datasources/SymptomLocalDataSource';
import { VitalSignRepositoryImpl } from '../data/repositories/VitalSignRepositoryImpl';
import { SymptomRepositoryImpl } from '../data/repositories/SymptomRepositoryImpl';
import { IVitalSignProvider } from '../domain/repositories/IVitalSignProvider';
import { ISymptomRepository } from '../domain/repositories/ISymptomRepository';
import { ObserveVitalSigns } from '../domain/usecases/ObserveVitalSigns';
import { CreateSymptomEntry } from '../domain/usecases/CreateSymptomEntry';
import { GetSymptomHistory } from '../domain/usecases/GetSymptomHistory';
import { DeleteSymptomEntry } from '../domain/usecases/DeleteSymptomEntry';
import { generateId } from '../shared/utils/idGenerator';

/**
 * Container de Injeção de Dependência (Composition Root).
 *
 * Centraliza a criação e wiring de todas as dependências da aplicação.
 * Este é o ÚNICO lugar onde implementações concretas são instanciadas
 * e conectadas às abstrações.
 *
 * Abordagem: Service Locator simples com instâncias singleton.
 *
 * Justificativa da escolha:
 * - Evita overhead de frameworks de DI (inversify, tsyringe) para o escopo atual
 * - Mantém a composição explícita e rastreável
 * - Permite troca fácil de implementações (ex: MockProvider → RealProvider)
 * - Pode migrar para Context-based DI no React se necessário
 *
 * Trade-off: menos automação que um framework de DI, porém mais transparência.
 */
export class ServiceContainer {
  // --- Data Sources (singleton) ---
  private readonly vitalSignDataSource: MockVitalSignProvider;
  private readonly symptomDataSource: SymptomLocalDataSource;

  // --- Repositories ---
  readonly vitalSignProvider: IVitalSignProvider;
  readonly symptomRepository: ISymptomRepository;

  // --- Use Cases ---
  readonly observeVitalSigns: ObserveVitalSigns;
  readonly createSymptomEntry: CreateSymptomEntry;
  readonly getSymptomHistory: GetSymptomHistory;
  readonly deleteSymptomEntry: DeleteSymptomEntry;

  constructor() {
    // 1. Instancia Data Sources
    this.vitalSignDataSource = new MockVitalSignProvider(3000);
    this.symptomDataSource = new SymptomLocalDataSource();

    // 2. Instancia Repositories (conectando interfaces a implementações)
    this.vitalSignProvider = new VitalSignRepositoryImpl(this.vitalSignDataSource);
    this.symptomRepository = new SymptomRepositoryImpl(this.symptomDataSource);

    // 3. Instancia Use Cases (injetando dependências do domínio)
    this.observeVitalSigns = new ObserveVitalSigns(this.vitalSignProvider);
    this.createSymptomEntry = new CreateSymptomEntry(this.symptomRepository, generateId);
    this.getSymptomHistory = new GetSymptomHistory(this.symptomRepository);
    this.deleteSymptomEntry = new DeleteSymptomEntry(this.symptomRepository);
  }
}

/**
 * Instância singleton do container.
 * Inicializada uma única vez no lifecycle do app.
 */
export const container = new ServiceContainer();

File: kotlin/app/src/main/java/com/vitalcheck/app/di/ServiceContainer.kt
package com.vitalcheck.app.di

import android.content.Context
import com.vitalcheck.app.data.datasource.MockVitalSignProvider
import com.vitalcheck.app.data.datasource.SymptomLocalDataSource
import com.vitalcheck.app.data.local.SymptomDatabase
import com.vitalcheck.app.data.repository.SymptomRepositoryImpl
import com.vitalcheck.app.data.repository.VitalSignRepositoryImpl
import com.vitalcheck.app.domain.repository.ISymptomRepository
import com.vitalcheck.app.domain.repository.IVitalSignProvider
import com.vitalcheck.app.domain.usecase.CreateSymptomEntry
import com.vitalcheck.app.domain.usecase.DeleteSymptomEntry
import com.vitalcheck.app.domain.usecase.GetSymptomHistory
import com.vitalcheck.app.domain.usecase.ObserveVitalSigns

/**
 * Composition Root — ponto único de criação e wiring de dependências.
 *
 * Equivalente ao ServiceContainer.ts da versão React Native.
 * Instancia todas as camadas respeitando o grafo de dependências
 * da Clean Architecture (domínio não depende de data/apresentação).
 *
 * Decisão: DI manual ao invés de Hilt/Dagger/Koin porque:
 * - 2 telas não justificam a complexidade de um framework DI
 * - Mais explícito e fácil de rastrear dependências
 * - Consistente com a abordagem da versão React Native
 * - Facilmente substituível por Hilt se o projeto crescer
 *
 * O container é instanciado uma vez no Application e exposto
 * via CompositionLocal para toda a árvore Compose.
 */
class ServiceContainer(context: Context) {

    // --- Data Sources ---
    private val database = SymptomDatabase.getInstance(context)
    private val symptomDao = database.symptomDao()
    private val vitalSignDataSource = MockVitalSignProvider(intervalMs = 3000L)
    private val symptomLocalDataSource = SymptomLocalDataSource(symptomDao)

    // --- Repositories ---
    private val vitalSignProvider: IVitalSignProvider =
        VitalSignRepositoryImpl(vitalSignDataSource)
    private val symptomRepository: ISymptomRepository =
        SymptomRepositoryImpl(symptomLocalDataSource)

    // --- Use Cases (expostos para ViewModels) ---
    val observeVitalSigns = ObserveVitalSigns(vitalSignProvider)
    val createSymptomEntry = CreateSymptomEntry(symptomRepository)
    val getSymptomHistory = GetSymptomHistory(symptomRepository)
    val deleteSymptomEntry = DeleteSymptomEntry(symptomRepository)
}

